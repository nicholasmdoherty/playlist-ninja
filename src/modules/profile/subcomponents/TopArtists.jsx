import React, { Component } from "react";
import { Image, Col, Row } from "react-bootstrap";
import { connect } from "react-redux";
import { deepCamelCaseKeys } from "../../../common/constants";
import Spacer from "../../../common/components/Spacer";

class TopArtists extends Component {
  constructor(props) {
    super(props);

    this.state = {
      topArtists: null,
      error: null
    };
  }

  componentDidMount() {
    this.props.api.spotifyApi.getMyTopArtists({ limit: 6 }).then(
      data => {
        this.setState({ topArtists: data.body });
      },
      error => {
        this.setState({ error: deepCamelCaseKeys(error) });
      }
    );

    this.forceUpdate();
  }

  render() {
    return (
      <Row className="p-3">
        <Col xs={12} className="text-center">
          <h3 className="display-5"> Top Artists </h3>
          <Spacer percentage={4} />
        </Col>

        {this.state.topArtists &&
          this.state.topArtists.items.map(topArtist => {
            return (
              <Col xs={6} sm={4} lg={2} className="text-center">
                <div className="p-2 break-long-words">
                  <Image
                    src={topArtist.images[0].url}
                    roundedCircle
                    fluid
                    className="drop-shadow mb-3"
                  />
                  <p className="lead">{topArtist.name}</p>
                </div>
              </Col>
            );
          })}
      </Row>
    );
  }
}

let mapStateToProps = state => {
  return {
    api: state.api
  };
};

export default connect(mapStateToProps)(TopArtists);
