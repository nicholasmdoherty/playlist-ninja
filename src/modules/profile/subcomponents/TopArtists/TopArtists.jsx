import React, { useEffect, useState } from "react";
import { connect } from "react-redux";

import './TopArtists.css'

const TopArtists = ({ api }) => {

  const [topArtists4Weeks, setTopArtists4Weeks] = useState([])
  const [topArtists6Months, setTopArtists6Months] = useState([])
  const [topArtistsAllTime, setTopArtistsAllTime] = useState([])
  const [activeView, setActiveView] = useState('short_term')
  const [error, setError] = useState(null)

  useEffect(() => {
    api.getMyTopArtists({ limit: 5, time_range: 'short_term' }).then(
      data => {
          setTopArtists4Weeks(data.body.items)
      },
      error => {
        setError(error)
      }
    );

    api.getMyTopArtists({ limit: 5, time_range: 'medium_term' }).then(
      data => {
          setTopArtists6Months(data.body.items)
      },
      error => {
        setError(error)
      }
    );

    api.getMyTopArtists({ limit: 5, time_range: 'long_term' }).then(
      data => {
          setTopArtistsAllTime(data.body.items)
      },
      error => {
        setError(error)
      }
    );
  }, [api])

  const renderTopArtists = () => {

    const renderArtist = (artist) => {
      return (
        <div className="artist-row">
          <img className='top-artist-image' width={40} height={40} src={artist.images.length > 0 ? artist.images[0].url : ''} />
          <span className="artist-name">{artist.name}</span>
        </div>
      )
    }

    if (error) {
      return 'Error loading top artists.'
    } else if (!topArtists4Weeks || !topArtists6Months || !topArtistsAllTime) {
      return 'Loading...'
    } else if (activeView === 'short_term') {
      return (
        <React.Fragment>
          {topArtists4Weeks.map((artist) => {
            return renderArtist(artist)
          })}
        </React.Fragment>
      )
    } else if (activeView === 'medium_term') {
      return (
        <React.Fragment>
          {topArtists6Months.map((artist) => {
            return renderArtist(artist)
          })}
        </React.Fragment>
      )
    } else if (activeView === 'long_term') {
      return (
        <React.Fragment>
          {topArtistsAllTime.map((artist) => {
            return renderArtist(artist)
          })}
        </React.Fragment>
      )
    }
  }

  return (
    <div className='top-artists-container'>
      <span className="top-artists-header">Top Artists</span>
      <div className='top-artists-button-row'>
        <div 
          className={`top-artists-button ${activeView === 'short_term' ? 'active' : ''}`}
          onClick={() => { setActiveView('short_term') }}
        >
          4 Weeks
        </div>
        <div 
          className={`top-artists-button ${activeView === 'medium_term' ? 'active' : ''}`}
          onClick={() => { setActiveView('medium_term') }}
        >
          6 Months
        </div>
        <div 
          className={`top-artists-button ${activeView === 'long_term' ? 'active' : ''}`}
          onClick={() => { setActiveView('long_term') }}
        >
          All Time
        </div>
      </div>
      <div className='card-style top-artists-card'>
        {renderTopArtists()}
      </div>
    </div>
  )
}

let mapStateToProps = state => {
  return {
    api: state.api.spotifyApi
  };
};

export default connect(mapStateToProps)(TopArtists);
