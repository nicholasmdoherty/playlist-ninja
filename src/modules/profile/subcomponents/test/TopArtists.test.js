// Package imports
import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";

// Project Imports
import TopArtists from "../TopArtists";
import configureStore from "../../../../store";

describe("TopArtists.jsx", () => {
  /**
   * Tests for the component with no top artists.
   *  - New spotify account, no data
   */
  describe("with empty artist array", () => {
    it("matches the snapshot", () => {
      const component = renderer.create(
        <Provider
          store={configureStore({
            api: {
              spotifyApi: {
                getMyTopArtists: () => {
                  return Promise.resolve({ body: { items: [] } });
                }
              }
            }
          })}
        >
          <TopArtists />
        </Provider>
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders without crashing", () => {
      const div = document.createElement("div");
      ReactDOM.render(
        <Provider
          store={configureStore({
            api: {
              spotifyApi: {
                getMyTopArtists: () => {
                  return Promise.resolve({ body: { items: [] } });
                }
              }
            }
          })}
        >
          <TopArtists />
        </Provider>,
        div
      );
      ReactDOM.unmountComponentAtNode(div);
    });
  });

  /**
   *  Tests for the component with top artist data
   */
  describe("recieved top artist data", () => {
    it("matches the snapshot", () => {
      const component = renderer.create(
        <Provider
          store={configureStore({
            api: {
              spotifyApi: {
                getMyTopArtists: () => {
                  return Promise.reject({
                    body: {
                      items: [
                        { images: null, name: "No Image" },
                        { images: [{ url: "fakeUrl" }], name: "Image" }
                      ]
                    }
                  });
                }
              }
            }
          })}
        >
          <TopArtists />
        </Provider>
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders without crashing", () => {
      const div = document.createElement("div");
      ReactDOM.render(
        <Provider
          store={configureStore({
            api: {
              spotifyApi: {
                getMyTopArtists: () => {
                  return Promise.reject({
                    body: {
                      items: [
                        { images: null, name: "No Image" },
                        { images: [{ url: "fakeUrl" }], name: "Image" }
                      ]
                    }
                  });
                }
              }
            }
          })}
        >
          <TopArtists />
        </Provider>,
        div
      );
      ReactDOM.unmountComponentAtNode(div);
    });
  });

  /**
   * Tests for the component with an error from Spotify.
   */
  describe("error from Spotify api", () => {
    it("matches the snapshot", () => {
      const component = renderer.create(
        <Provider
          store={configureStore({
            api: {
              spotifyApi: {
                getMyTopArtists: () => {
                  return Promise.reject("error");
                }
              }
            }
          })}
        >
          <TopArtists />
        </Provider>
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders without crashing", () => {
      const div = document.createElement("div");
      ReactDOM.render(
        <Provider
          store={configureStore({
            api: {
              spotifyApi: {
                getMyTopArtists: () => {
                  return Promise.reject("error");
                }
              }
            }
          })}
        >
          <TopArtists />
        </Provider>,
        div
      );
      ReactDOM.unmountComponentAtNode(div);
    });
  });
});
