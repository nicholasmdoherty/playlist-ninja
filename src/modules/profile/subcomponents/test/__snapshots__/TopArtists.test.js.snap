// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopArtists.jsx error from Spotify api matches the snapshot 1`] = `
<div
  className="p-3 row"
>
  <div
    className="text-center col-12"
  >
    <h3
      className="display-5"
    >
       Top Artists 
    </h3>
    <div
      style={
        Object {
          "height": "4vh",
        }
      }
    />
  </div>
</div>
`;

exports[`TopArtists.jsx recieved top artist data matches the snapshot 1`] = `
<div
  className="p-3 row"
>
  <div
    className="text-center col-12"
  >
    <h3
      className="display-5"
    >
       Top Artists 
    </h3>
    <div
      style={
        Object {
          "height": "4vh",
        }
      }
    />
  </div>
</div>
`;

exports[`TopArtists.jsx with empty artist array matches the snapshot 1`] = `
<div
  className="p-3 row"
>
  <div
    className="text-center col-12"
  >
    <h3
      className="display-5"
    >
       Top Artists 
    </h3>
    <div
      style={
        Object {
          "height": "4vh",
        }
      }
    />
  </div>
</div>
`;
